File: packages/ios-swift/Sources/CinacoinSDK/CinacoinSDK.swift
//
//  CinacoinSDK.swift
//  CinacoinSDK
//
//  Main entry point — singleton facade for the Cinacoin iOS Wallet SDK.
//
//  Usage:
//  ```swift
//  import CinacoinSDK
//
//  let sdk = CinacoinSDK.shared
//
//  try await sdk.configure(
//      projectId: "YOUR_PROJECT_ID",
//      metadata: AppMetadata(
//          name: "My dApp",
//          description: "An iOS dApp using Cinacoin",
//          url: "https://mydapp.com",
//          icons: ["https://mydapp.com/icon.png"]
//      )
//  )
//
//  let result = try await sdk.connect(walletId: "metamask")
//  print("Connected: \(result.accounts)")
//
//  let sig = try await sdk.signMessage("Hello from Cinacoin!")
//  try await sdk.disconnect()
//  ```
//

import Foundation
import Combine

// MARK: - CinacoinSDK (Singleton)

/// The top-level Cinacoin Wallet SDK.
///
/// `CinacoinSDK.shared` provides a thread-safe, singleton entry point
/// that coordinates wallet connection, signing, transactions, chain
/// switching, and session persistence.
///
/// All async operations use Swift Concurrency (`async/await`).
/// Published properties use Combine for reactive SwiftUI bindings.
@MainActor
public final class CinacoinSDK: ObservableObject {

    // MARK: - Shared Instance

    /// The global shared SDK instance.
    public static let shared = CinacoinSDK()

    // MARK: - Published State (SwiftUI)

    @Published public private(set) var status: ConnectionStatus = .disconnected
    @Published public private(set) var accounts: [String] = []
    @Published public private(set) var chainId: Int = 1

    /// Convenience: whether a wallet is connected.
    public var isConnected: Bool { status == .connected }

    // MARK: - Sub-Managers

    /// WalletConnect pairing and session orchestration.
    public let walletConnect = WalletConnectManager()

    /// EVM chain registry and active-chain management.
    public let chainManager = ChainManager()

    /// Message signing (personal_sign, signTypedData).
    public private(set) lazy var signer = SignerManager(walletConnect: walletConnect)

    /// Transaction building and sending.
    public private(set) lazy var transactions = TransactionManager(
        walletConnect: walletConnect,
        chainManager: chainManager
    )

    /// Network reachability monitor.
    public let networkMonitor = NetworkMonitor()

    /// Keychain-backed session storage.
    public let storage = StorageManager()

    // MARK: - Events

    private let eventSubject = PassthroughSubject<CinacoinEvent, Never>()

    /// Combine publisher for SDK events.
    public var events: AnyPublisher<CinacoinEvent, Never> {
        eventSubject.eraseToAnyPublisher()
    }

    // MARK: - Internal State

    private var _configured = false
    private var _disposed = false
    private let lock = NSRecursiveLock()

    // MARK: - Private Init

    private init() {}

    // MARK: - Configure

    /// Initialize the SDK. Must be called once before any other method.
    ///
    /// 1. Configures WalletConnect v2.
    /// 2. Registers the chain manager with supported EVM chains.
    /// 3. Attempts session restoration from the Keychain.
    /// 4. Starts network monitoring.
    ///
    /// - Parameters:
    ///   - projectId: WalletConnect / Cinacoin project ID.
    ///   - metadata:  App metadata for wallet pairing.
    ///   - relayUrl:  Optional WebSocket relay override.
    ///   - namespace: Default chain namespace (default `"eip155"`).
    public func configure(
        projectId: String,
        metadata: AppMetadata,
        relayUrl: String? = nil,
        namespace: String = "eip155"
    ) async throws {
        try _guardNotDisposed()
        lock.lock()
        defer { lock.unlock() }
        guard !_configured else { return }

        // Configure WalletConnect.
        try await walletConnect.configure(
            projectId: projectId,
            metadata: metadata,
            relayUrl: relayUrl
        )

        // Attempt session restoration from Keychain.
        do {
            if let session = try storage.loadSession() {
                accounts = session.accounts
                chainId = session.chainId
                status = .connected
                signer.setPrimaryAddress(session.accounts.first ?? "")
                transactions.setPrimaryAddress(session.accounts.first ?? "")
                eventSubject.send(.sessionRestored(
                    accounts: session.accounts,
                    chainId: session.chainId
                ))
            }
        } catch {
            // Keychain error is non-fatal; proceed without restoration.
        }

        // Start network monitoring.
        networkMonitor.start()

        _configured = true
        eventSubject.send(.configured)
    }

    // MARK: - Connect

    /// Connect to a wallet.
    ///
    /// Creates a WalletConnect pairing, opens the wallet app via deep link,
    /// and waits for session approval.
    ///
    /// - Parameters:
    ///   - walletId: Connector ID (e.g. `"metamask"`, `"walletconnect"`).
    ///   - chains:   Optional chain IDs to request.
    /// - Returns: `ConnectionResult` with session info and accounts.
    public func connect(walletId: String, chains: [Int]? = nil) async throws -> ConnectionResult {
        try _guardConfigured()

        status = .connecting
        eventSubject.send(.connecting(connectorId: walletId))

        let result = try await walletConnect.connect(walletId: walletId, chains: chains)

        // Persist session.
        let session = SessionData(
            accounts: result.accounts,
            chainId: result.chainId,
            sessionId: result.sessionId,
            connectorId: result.connectorId,
            topic: result.sessionId
        )
        try? storage.saveSession(session)

        // Update state.
        accounts = result.accounts
        chainId = result.chainId
        status = .connected
        signer.setPrimaryAddress(result.accounts.first ?? "")
        transactions.setPrimaryAddress(result.accounts.first ?? "")

        eventSubject.send(.connected(accounts: result.accounts, chainId: result.chainId))

        return result
    }

    // MARK: - Disconnect

    /// Disconnect the active wallet session.
    public func disconnect() async {
        do { try _guardConfigured() } catch { return }

        await walletConnect.disconnect()
        try? storage.clearSession()

        accounts = []
        chainId = 1
        status = .disconnected

        eventSubject.send(.disconnected)
    }

    // MARK: - Chain Switching

    /// Switch the active EVM chain.
    ///
    /// - Parameter chainId: Target chain ID.
    /// - Throws: `CinacoinError.chainNotSupported` if the chain is not registered.
    public func switchChain(chainId: Int) async throws {
        try _guardConfigured()
        guard isConnected else { throw CinacoinError.notConnected }

        try chainManager.setActiveChain(chainId: chainId)
        try await walletConnect.switchChain(chainId)

        // Update persisted session.
        if let session = try? storage.loadSession() {
            let updated = SessionData(
                accounts: session.accounts,
                chainId: chainId,
                sessionId: session.sessionId,
                connectorId: session.connectorId,
                topic: session.topic
            )
            try? storage.saveSession(updated)
        }

        self.chainId = chainId
        eventSubject.send(.chainChanged(chainId: chainId))
    }

    // MARK: - Signing

    /// Sign a plaintext message (EIP-191 `personal_sign`).
    ///
    /// - Parameter message: The plaintext message to sign.
    /// - Returns: `SignatureResult` with hex-encoded signature.
    public func signMessage(_ message: String) async throws -> SignatureResult {
        try _guardConfigured()
        guard isConnected else { throw CinacoinError.notConnected }
        return try await signer.personalSign(message: message)
    }

    /// Sign EIP-712 typed data.
    ///
    /// - Parameter typedDataJson: JSON-encoded EIP-712 typed data.
    /// - Returns: `SignatureResult` with hex-encoded signature.
    public func signTypedData(_ typedDataJson: String) async throws -> SignatureResult {
        try _guardConfigured()
        guard isConnected else { throw CinacoinError.notConnected }
        return try await signer.signTypedData(typedDataJson: typedDataJson)
    }

    // MARK: - Transactions

    /// Send a transaction through the connected wallet.
    ///
    /// - Parameter tx: Transaction parameters.
    /// - Returns: Transaction hash (`0x`-prefixed hex string).
    public func sendTransaction(_ tx: TransactionRequest) async throws -> String {
        try _guardConfigured()
        guard isConnected else { throw CinacoinError.notConnected }
        return try await transactions.sendTransaction(tx)
    }

    /// Estimate gas for a transaction.
    public func estimateGas(_ tx: TransactionRequest) async throws -> String {
        try _guardConfigured()
        return try await transactions.estimateGas(tx)
    }

    // MARK: - Balance

    /// Fetch the native token balance for the primary connected account.
    ///
    /// - Returns: Balance in wei as a hex string.
    public func getBalance() async throws -> String {
        try _guardConfigured()
        guard isConnected else { throw CinacoinError.notConnected }
        guard let address = accounts.first else { throw CinacoinError.notConnected }
        return try await transactions.getBalance(address: address, chainId: chainId)
    }

    // MARK: - SIWE

    /// Generate a Sign-In With Ethereum (EIP-4361) message.
    ///
    /// - Parameters:
    ///   - domain:    The dApp domain (e.g. `"mydapp.com"`).
    ///   - nonce:     Server-generated nonce for replay protection.
    ///   - uri:       Optional URI override.
    ///   - statement: Optional human-readable statement.
    /// - Returns: SIWE message string ready for signing.
    public func generateSiweMessage(
        domain: String,
        nonce: String,
        uri: String? = nil,
        statement: String? = nil
    ) throws -> String {
        guard isConnected else { throw CinacoinError.notConnected }
        guard let address = accounts.first else { throw CinacoinError.notConnected }

        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let now = formatter.string(from: Date())

        var lines: [String] = []
        lines.append("\(domain) wants you to sign in with your Ethereum account:")
        lines.append(address)
        lines.append("")
        if let statement = statement {
            lines.append(statement)
            lines.append("")
        }
        lines.append("URI: \(uri ?? "https://\(domain)")")
        lines.append("Version: 1")
        lines.append("Chain ID: \(chainId)")
        lines.append("Nonce: \(nonce)")
        lines.append("Issued At: \(now)")

        return lines.joined(separator: "\n")
    }

    // MARK: - Utilities

    /// Get a WalletConnect pairing URI for QR display.
    public func getPairingUri() async throws -> String? {
        try _guardConfigured()
        return try await walletConnect.createPairing()
    }

    /// Check if a wallet app is installed on the device.
    public func isWalletInstalled(_ walletId: String) -> Bool {
        walletConnect.isWalletInstalled(walletId)
    }

    /// Get a list of recommended wallets.
    public func getRecommendedWallets() -> [WalletInfo] {
        WalletRegistry.shared.recommended()
    }

    /// Get a block explorer URL for a transaction.
    public func explorerUrl(for txHash: String) -> String? {
        transactions.explorerUrl(for: txHash)
    }

    // MARK: - Disposal

    /// Dispose of all resources and reset state.
    public func dispose() async {
        lock.lock()
        defer { lock.unlock() }
        guard !_disposed else { return }
        _disposed = true

        networkMonitor.stop()
        await walletConnect.dispose()
        try? storage.clearSession()

        accounts = []
        chainId = 1
        status = .disconnected
        _configured = false
    }

    // MARK: - Guards

    private func _guardConfigured() throws {
        guard _configured else { throw CinacoinError.notInitialized }
        guard !_disposed else { throw CinacoinError.disposed }
    }

    private func _guardNotDisposed() throws {
        guard !_disposed else { throw CinacoinError.disposed }
    }
}

// MARK: - WalletRegistry

/// Static registry of known wallet apps for discovery and deep-linking.
public final class WalletRegistry: @unchecked Sendable {
    public static let shared = WalletRegistry()

    private let wallets: [WalletInfo] = [
        WalletInfo(
            id: "metamask", name: "MetaMask",
            deepLinkScheme: "metamask://",
            universalLinkDomain: "metamask.app.link",
            appStoreUrl: "https://apps.apple.com/app/id1438144202",
            supportedChains: ["eip155:1", "eip155:137", "eip155:42161", "eip155:10", "eip155:56"]
        ),
        WalletInfo(
            id: "walletconnect", name: "WalletConnect",
            supportedChains: ["eip155:1", "eip155:137"]
        ),
        WalletInfo(
            id: "coinbase", name: "Coinbase Wallet",
            deepLinkScheme: "cbwallet://",
            universalLinkDomain: "go.cb-w.com",
            appStoreUrl: "https://apps.apple.com/app/id1278383455",
            supportedChains: ["eip155:1", "eip155:10", "eip155:8453"]
        ),
        WalletInfo(
            id: "rainbow", name: "Rainbow",
            deepLinkScheme: "rainbow://",
            appStoreUrl: "https://apps.apple.com/app/id1457119021",
            supportedChains: ["eip155:1", "eip155:10", "eip155:42161", "eip155:8453"]
        ),
        WalletInfo(
            id: "trust", name: "Trust Wallet",
            deepLinkScheme: "trust://",
            appStoreUrl: "https://apps.apple.com/app/id1288339409",
            supportedChains: ["eip155:1", "eip155:56", "eip155:137"]
        ),
        WalletInfo(
            id: "phantom", name: "Phantom",
            deepLinkScheme: "phantom://",
            appStoreUrl: "https://apps.apple.com/app/id1598432977",
            supportedChains: ["eip155:1", "eip155:137", "eip155:42161"]
        ),
    ]

    /// Return all recommended wallets (filtered for EVM chains).
    public func recommended() -> [WalletInfo] {
        wallets.filter { w in
            w.supportedChains.contains { $0.hasPrefix("eip155") }
        }
    }

    /// Look up a wallet by ID.
    public func wallet(id: String) -> WalletInfo? {
        wallets.first { $0.id == id }
    }
}

File: onux/packages/flutter-dart/lib/src/cinacoin_sdk.dart
/// Cinacoin SDK main entry point.
///
/// Singleton facade coordinating WalletConnect, chain adapters,
/// transaction management, and signing.
library cinacoin_sdk;

import 'dart:async';

import 'walletconnect_manager.dart';
import 'chain_adapters.dart';
import 'transaction_manager.dart';
import 'signer_manager.dart';
import 'storage_manager.dart';
import 'types.dart';

/// Main SDK singleton facade.
///
/// Provides a unified API for:
/// - WalletConnect v2 pairing and session management
/// - Multi-chain adapter support (12 chains)
/// - Transaction building, signing, and polling
/// - Message signing (EIP-191, EIP-712)
/// - Encrypted session persistence
///
/// ### Usage
///
/// ```dart
/// // Initialize
/// final sdk = CinacoinSdk.instance;
/// await sdk.initialize(
///   projectId: 'YOUR_PROJECT_ID',
///   metadata: AppMetadata(
///     name: 'My dApp',
///     description: 'A Flutter dApp',
///     url: 'https://mydapp.com',
///     icons: [],
///   ),
/// );
///
/// // Connect
/// final pairing = await sdk.createPairing();
/// // Display pairing.uri as QR code
///
/// // After wallet approves
/// final session = await sdk.getSession();
///
/// // Sign
/// final sig = await sdk.personalSign(message: 'Hello!');
///
/// // Send transaction
/// final txHash = await sdk.sendViaWalletConnect(
///   from: accounts[0],
///   to: '0x...',
///   value: '1000000000000000000',
/// );
///
/// // Switch chain
/// await sdk.switchChain('eip155:137');
///
/// // Disconnect
/// await sdk.disconnect();
/// ```
class CinacoinSdk {
  /// Singleton instance.
  static final CinacoinSdk instance = CinacoinSdk._();

  CinacoinSdk._();

  // ── Internal State ───────────────────────────────────────────────────

  WalletConnectManager? _wcManager;
  TransactionManager? _txManager;
  SignerManager? _signerManager;
  StorageManager? _storageManager;
  bool _initialized = false;
  bool _disposed = false;

  final _stateController = StreamController<SdkState>.broadcast();

  /// Stream of SDK state changes.
  Stream<SdkState> get stateChanges => _stateController.stream;

  /// Whether the SDK is initialized.
  bool get isInitialized => _initialized;

  /// Whether a wallet session is active.
  bool get isConnected => _wcManager?.hasActiveSession ?? false;

  /// Active session topic.
  String? get sessionTopic => _wcManager?.activeSessionTopic;

  /// Connected accounts.
  List<String> get accounts {
    final session = _wcManager?.activeSession;
    if (session == null) return [];
    final accounts = <String>[];
    for (final ns in session.namespaces.values) {
      for (final account in ns.accounts) {
        final parts = account.split(':');
        if (parts.length == 3) accounts.add(parts[2]);
      }
    }
    return accounts;
  }

  // ── Initialization ───────────────────────────────────────────────────

  /// Initialize the SDK.
  ///
  /// [projectId] WalletConnect project ID.
  /// [metadata] App metadata for pairing.
  /// [relayUrl] Optional relay URL override.
  /// [chains] Chain IDs to register adapters for (default: all).
  /// [sessionTtl] Session persistence TTL.
  Future<void> initialize({
    required String projectId,
    required AppMetadata metadata,
    String? relayUrl,
    List<String>? chains,
    Duration sessionTtl = const Duration(days: 7),
  }) async {
    if (_initialized || _disposed) return;

    // Initialize storage
    _storageManager = StorageManager(sessionTtl: sessionTtl);
    await _storageManager!.init();

    // Initialize WalletConnect manager
    _wcManager = WalletConnectManager(
      projectId: projectId,
      metadata: metadata,
      relayUrl: relayUrl,
      sessionTtl: sessionTtl,
    );
    await _wcManager!.init();

    // Initialize transaction manager
    _txManager = TransactionManager();

    // Register chain adapters
    final targetChains = chains ?? ChainRegistry.getAllChainIds();
    for (final chainId in targetChains) {
      final config = ChainRegistry.get(chainId);
      if (config != null) {
        _txManager!.registerAdapter(chainId, ChainRegistry.getAdapter(chainId));
      }
    }

    // Initialize signer manager
    _signerManager = SignerManager(
      requestFn: ({
        required String method,
        required List<dynamic> params,
        String? topic,
        String? chainId,
      }) =>
          _wcManager!.request(
        method: method,
        params: params,
        topic: topic,
        chainId: chainId,
      ),
      sessionTopicGetter: () => _wcManager!.activeSessionTopic,
      chainIdGetter: () => 1,
      accountsGetter: () => accounts,
    );

    _initialized = true;
    _emitState();

    // Attempt session restore
    await _wcManager!.tryRestoreSession();
  }

  // ── Pairing ──────────────────────────────────────────────────────────

  /// Create a new WalletConnect pairing.
  ///
  /// Returns a PairingResult with the URI for QR display.
  Future<PairingResult> createPairing({
    List<String> chains = const ['eip155:1'],
    List<String>? methods,
    List<String>? events,
  }) async {
    _throwIfNotInitialized();
    return _wcManager!.createPairing(
      chains: chains,
      methods: methods,
      events: events,
    );
  }

  /// Pair using a WC URI.
  Future<void> pairFromUri(String uri) async {
    _throwIfNotInitialized();
    await _wcManager!.pairFromUri(uri);
  }

  // ── Session ──────────────────────────────────────────────────────────

  /// Get the active session.
  SessionStruct? getSession() => _wcManager?.activeSession;

  /// Approve a session proposal.
  Future<WcSessionResult> approveSession({
    required String proposerPublicKey,
    required Map<String, RequiredNamespace> namespaces,
  }) async {
    _throwIfNotInitialized();
    return _wcManager!.approveSession(
      proposerPublicKey: proposerPublicKey,
      namespaces: namespaces,
    );
  }

  /// Disconnect the active session.
  Future<void> disconnect() async {
    _throwIfNotInitialized();
    await _wcManager!.disconnect();
    _emitState();
  }

  // ── Chain Switching ──────────────────────────────────────────────────

  /// Switch the active chain.
  Future<void> switchChain(String chainId) async {
    _throwIfNotInitialized();
    _throwIfNotConnected();

    if (!ChainRegistry.isSupported(chainId)) {
      throw ArgumentError('Unsupported chain: $chainId');
    }

    _txManager!.setActiveChain(chainId);
    await _storageManager?.setLastChainId(int.parse(chainId.split(':').last));
    _emitState();
  }

  /// Get the last used chain ID.
  Future<int?> getLastChainId() async {
    await _storageManager?.init();
    return _storageManager?.getLastChainId();
  }

  // ── Signing ──────────────────────────────────────────────────────────

  /// Sign a message using EIP-191 personal_sign.
  Future<SignatureResult> personalSign({
    required String message,
    String? chainId,
  }) async {
    _throwIfNotInitialized();
    _throwIfNotConnected();
    return _signerManager!.personalSign(message: message, chainId: chainId);
  }

  /// Sign EIP-712 typed data (v4).
  Future<SignatureResult> signTypedDataV4({
    required String typedData,
    String? chainId,
  }) async {
    _throwIfNotInitialized();
    _throwIfNotConnected();
    return _signerManager!.signTypedDataV4(typedData: typedData, chainId: chainId);
  }

  // ── Transactions ─────────────────────────────────────────────────────

  /// Send a transaction via the connected wallet.
  Future<String> sendViaWalletConnect({
    required String from,
    required String to,
    String? value,
    String? data,
    String? chainId,
  }) async {
    _throwIfNotInitialized();
    _throwIfNotConnected();

    return _txManager!.sendViaWalletConnect(
      from: from,
      to: to,
      value: value,
      data: data,
      chainId: chainId,
      request: ({
        required String method,
        required List<dynamic> params,
        String? topic,
        String? cId,
      }) =>
          _wcManager!.request(
        method: method,
        params: params,
        topic: topic,
        chainId: cId,
      ),
    );
  }

  /// Estimate fees for a transaction.
  Future<String> estimateFee({
    required String from,
    required String to,
    String? value,
    String? data,
    String? chainId,
  }) async {
    _throwIfNotInitialized();
    return _txManager!.estimateFee(
      from: from,
      to: to,
      value: value,
      data: data,
      chainId: chainId,
    );
  }

  /// Poll a transaction for confirmation.
  Future<TxEvent> pollTransaction(
    String txHash, {
    int maxAttempts = 30,
    int intervalMs = 3000,
    String? chainId,
  }) async {
    _throwIfNotInitialized();
    return _txManager!.pollTransaction(
      txHash,
      maxAttempts: maxAttempts,
      intervalMs: intervalMs,
      chainId: chainId,
    );
  }

  /// Get balance for an address.
  Future<String> getBalance(String address, {String? chainId}) async {
    _throwIfNotInitialized();
    return _txManager!.getBalance(address, chainId: chainId);
  }

  /// Get the latest block.
  Future<String> getLatestBlock({String? chainId}) async {
    _throwIfNotInitialized();
    return _txManager!.getLatestBlock(chainId: chainId);
  }

  // ── Storage ──────────────────────────────────────────────────────────

  /// Get the storage manager for custom data persistence.
  StorageManager? get storage => _storageManager;

  // ── Utilities ────────────────────────────────────────────────────────

  /// Get the pairing URI for QR display.
  String? getPairingUri() => _wcManager?.getPairingUri();

  /// Get all supported chain IDs.
  List<String> getSupportedChains() => ChainRegistry.getAllChainIds();

  /// Get chain config by ID.
  ChainConfig? getChainConfig(String chainId) => ChainRegistry.get(chainId);

  // ── Disposal ─────────────────────────────────────────────────────────

  /// Dispose of all SDK resources.
  Future<void> dispose() async {
    if (_disposed) return;
    _disposed = true;

    await _txManager?.dispose();
    await _wcManager?.dispose();
    await _stateController.close();
    _initialized = false;
  }

  // ── Internal ─────────────────────────────────────────────────────────

  void _emitState() {
    _stateController.add(SdkState(
      isInitialized: _initialized,
      isConnected: isConnected,
      accounts: accounts,
      sessionTopic: sessionTopic,
    ));
  }

  void _throwIfNotInitialized() {
    if (!_initialized) {
      throw StateError('CinacoinSdk not initialized. Call initialize() first.');
    }
    if (_disposed) {
      throw StateError('CinacoinSdk has been disposed.');
    }
  }

  void _throwIfNotConnected() {
    if (!isConnected) {
      throw StateError('No wallet connected. Call createPairing() and approve a session first.');
    }
  }
}

/// SDK state snapshot.
class SdkState {
  final bool isInitialized;
  final bool isConnected;
  final List<String> accounts;
  final String? sessionTopic;

  const SdkState({
    required this.isInitialized,
    required this.isConnected,
    required this.accounts,
    this.sessionTopic,
  });

  @override
  String toString() =>
      'SdkState(initialized=$isInitialized, connected=$isConnected, accounts=$accounts)';
}
